sycl/doc/design/VirtualFunctions.md

```diff
@@ -0,0 +1,428 @@
+  // kernel, regardless of whether a target device supports 'fp64' aspect which
+  // is used by 'Foo::foo'.
+  // No exceptions are expected to be thrown.
+  new (Storage) Foo;
+});
+
+if (q.get_device().has(sycl::aspect::fp64)) {
+  auto props = syclext::properties{syclext::calls_indirectly<fp64_set>};
+  q.single_task(props, [=] {
+    Storage->foo();
+  });
+} else {
+  auto props = syclext::properties{syclext::calls_indirectly<regular_set>};
+  q.single_task(props, [=] {
+    Storage->bar();
+  });
+}
+```
+
+This example should work regardless of whether target device supports 'fp64'
+aspect or not. Implementation differs for JIT and AOT flows.
+
+##### JIT flow
+
+Regardless of device code split mode selected by a user, functions marked with
+`indirectly_callable` property should be outlined into separate device images
+by `sycl-post-link` tool based on the argument of the `indirectly_callable`
+property, i.e. all functions from the same set should be bundled into a
+dedicated device image.
+
+**TODO**: as an optimization, we can consider preserving virtual functions from
+sets that do not use any optional kernel features.
+
+Virtual functions in the original device image should be turned into
+declarations instead of definitions.
+
+Additionally, if any virtual function in such device image uses any optional
+kernel features, then the whole image should be cloned with all function bodies
+emptied. This cloned device image will be further referred to as "dummy virtual
+functions device image".
+
+This dummy device image is needed to support the example showed above when a
+kernel creates an object of a polymorphic class where some of virtual functions
+use optional features. LLVM IR generated by front-end will contain a vtable,
+which references all methods of the class. However, not all of them can be
+directly included into kernel's device image to avoid speculative compilation.
+
+When such kernel is submitted to a device, runtime will check which optional
+features are supported and link one or another device image with virtual
+functions.
+
+##### AOT flow
+
+In AOT mode, there will be no dynamic linking, but at the same time we know the
+list of supported optional features by a device thanks to
+[device config file][device-config-file-design].
+
+Therefore, `sycl-post-link` should read the device config file to determine list
+of optional features supported by a target and based on that drop all virtual
+functions from sets that use unsupported optional features.
+
+Note that we are making decision not based on which aspects are used by each
+individual virtual functions, but based on which aspects are used by a set of
+virtual functions (as identified by the `indirectlly_callable` property
+argument). The latter is computed as conjunction of aspects used by each
+virtual function within a set.
+
+The behavior is defined this way to better match the extension speficiation
```

Comment:
```suggestion
The behavior is defined this way to better match the extension specification
```

Reply:
Applied in 683bc66e92e3acb4c652791cfd7a815e5afa5783